collector/netframework_clr.go

```
@@ -0,0 +1,1218 @@
+		"Regexp of processes to include. Process name must both match whitelist and not match blacklist to be included.",
+	).Default(".*").String()
+	netclrBlacklist = kingpin.Flag(
+		"collector.netframework_clr.blacklist",
+		"Regexp of processes to exclude. Process name must both match whitelist and not match blacklist to be included.",
+	).Default("").String()
+)
+
+type netclrCollectorFunc func(ctx *ScrapeContext, ch chan<- prometheus.Metric) (*prometheus.Desc, error)
+
+type netclrCollectorsMap map[string]netclrCollectorFunc
+
+func (c *netclrCollector) getNETCLRCollectors() netclrCollectorsMap {
+	collectors := make(netclrCollectorsMap)
+	collectors["exceptions"] = c.collectExceptions
+	collectors["interop"] = c.collectInterop
+	collectors["jit"] = c.collectJit
+	collectors["loading"] = c.collectLoading
+	collectors["locksandthreads"] = c.collectLocksAndThreads
+	collectors["memory"] = c.collectMemory
+	collectors["remoting"] = c.collectRemoting
+	collectors["security"] = c.collectSecurity
+
+	return collectors
+}
+
+func netclrAvailableCollectors() string {
+	return "exceptions,interop,jit,loading,locksandthreads,memory,remoting,security"
+}
+
+func netclrExpandEnabledCollectors(enabled string) []string {
+	separated := strings.Split(enabled, ",")
+	unique := map[string]bool{}
+	for _, s := range separated {
+		if s != "" {
+			unique[s] = true
+		}
+	}
+	result := make([]string, 0, len(unique))
+	for s := range unique {
+		result = append(result, s)
+	}
+	return result
+}
+
+func netclrGetPerfObjectName(collector string) string {
+	switch collector {
+	case "exceptions":
+		return ".NET CLR Exceptions"
+	case "interop":
+		return ".NET CLR Interop"
+	case "jit":
+		return ".NET CLR Jit"
+	case "loading":
+		return ".NET CLR Loading"
+	case "locksandthreads":
+		return ".NET CLR LocksAndThreads"
+	case "memory":
+		return ".NET CLR Memory"
+	case "remoting":
+		return ".NET CLR Remoting"
+	case "security":
+		return ".NET CLR Security"
+	default:
+		return ""
```

We should probably fail for invalid configuration here?

---

Yes, I've set the function to return `(string, error)` and handled that with each call.
`default` case now returns an error.